File: src/card/history-card.ts
import type { HistoryTool, HistoryTurn, ThreadHistory } from '../agent/types';
import {
  card,
  collapsiblePanel,
  collapsiblePanelEl,
  hr,
  md,
  note,
  noteMd,
  type CardElement,
  type CardObject,
} from './cards';
import { relativeTime } from './command-cards';

/**
 * The "已恢复历史会话" card — posted as the first message of the resumed topic
 * (reply_in_thread on the /resume message creates the topic). It is a pure,
 * non-streaming digest: every turn is a collapsed {@link collapsiblePanel}
 * (collapse/expand is client-side, no callback), and reasoning + tool detail
 * fold one level deeper ("一层层下钻"). The codex session itself is already
 * bound to this topic and resumes on the next message — this card is just the
 * "where we left off" you see before continuing.
 */
export interface HistoryCardState {
  cwd: string;
  projectName?: string;
  history: ThreadHistory;
}

// Per-field truncation — Feishu 400s a card if any element nears ~30KB, so the
// transcript is a digest, not a full replay (codex keeps the real context).
const USER_MAX = 300;
const ASSIST_MAX = 800;
const REASON_MAX = 600;
const TOOL_TITLE_MAX = 90;
const TOOLS_BODY_MAX = 700;
const TOOLS_MAX_LINES = 12;
const PREVIEW_MAX = 160;
// Collapsed-state title = the user's question, one line.
const TITLE_Q_MAX = 56;
// JSON overhead of a collapsible_panel wrapper (header/icon/border/padding),
// charged on top of its body so the budget reflects real serialized size.
const PANEL_SHELL = 360;
// Whole-card content budget. Set well under Feishu's ~30KB hard limit to leave
// headroom for the header / meta / footer / preview that aren't counted in
// `used`; older turns past it degrade to a title-only panel.
const BODY_BUDGET = 18_000;

export function buildHistoryCard(state: HistoryCardState): CardObject {
  const { history } = state;
  const elements: CardElement[] = [metaNote(state)];

  if (history.turns.length === 0) {
    elements.push(
      hr(),
      md('_这个会话还没有可显示的历史（可能是空会话或刚创建）。_'),
      hr(),
      resumedFooter(),
    );
    return card(elements, { header: header(state), summary: '已恢复历史会话' });
  }

  const dropped = history.totalTurns - history.turns.length;
  if (dropped > 0) {
    elements.push(note(`仅显示最近 ${history.turns.length} 轮，更早的 ${dropped} 轮 Codex 仍保留在上下文中。`));
  }
  elements.push(hr());

  // Allocate budget newest→oldest so the most recent turns stay full and only
  // the oldest degrade to a title-only panel; emit oldest→newest for reading.
  const panels: CardElement[] = [];
  let used = 0;
  for (let i = history.turns.length - 1; i >= 0; i--) {
    const turn = history.turns[i];
    if (!turn) continue;
    const title = turnTitle(turn);
    const body = turnBody(turn);
    const size = estimateSize(body) + PANEL_SHELL;
    if (used + size > BODY_BUDGET && panels.length > 0) {
      const stub = '_（内容已省略，历史较长）_';
      panels.push(collapsiblePanel({ title, expanded: false, border: 'grey', body: stub }));
      used += title.length + stub.length + PANEL_SHELL;
    } else {
      panels.push(collapsiblePanelEl({ title, expanded: false, border: 'grey', elements: body }));
      used += size;
    }
  }
  panels.reverse();
  elements.push(...panels);

  const last = history.turns[history.turns.length - 1];
  const leftOff = last ? last.assistantText || last.userText || (last.tools.at(-1)?.title ?? '') : '';
  if (leftOff.trim()) {
    elements.push(hr(), noteMd(`📍 **上次停在**：${truncateTail(leftOff, PREVIEW_MAX)}`));
  } else {
    elements.push(hr());
  }
  elements.push(resumedFooter());

  return card(elements, { header: header(state), summary: `已恢复历史会话 · 共 ${history.totalTurns} 轮` });
}

function header(state: HistoryCardState): { title: string; template: 'turquoise'; subtitle?: string } {
  const { history } = state;
  const bits = [
    history.name?.trim() || state.projectName,
    `共 ${history.totalTurns} 轮`,
    history.updatedAt ? relativeTime(history.updatedAt) : undefined,
  ].filter(Boolean);
  return { title: '🕘 已恢复历史会话', template: 'turquoise', subtitle: bits.join(' · ') };
}

function metaNote(state: HistoryCardState): CardElement {
  const parts = [`📂 \`${state.cwd}\``];
  if (state.projectName) parts.unshift(`📁 ${state.projectName}`);
  return note(parts.join('   '));
}

function resumedFooter(): CardElement {
  return md('✅ **会话已恢复** —— 直接发消息即可继续。');
}

/**
 * Collapsed-state title of a turn: the user's question on one line, with a
 * trailing '…' when there's more. Tapping expands to the full Q&A + the
 * 思考/工具 detail fold (see {@link turnBody}) — the answer lives there, not in
 * the title, so the collapsed list stays a clean scan of the questions.
 */
function turnTitle(turn: HistoryTurn): string {
  if (turn.userText.trim()) return `👤 ${escapeInline(truncate(oneLine(turn.userText), TITLE_Q_MAX))}`;
  return '⚙️ 系统 / 工具调用';
}

/** The expandable body of one turn: full Q&A + a deeper fold for 思考/工具. */
function turnBody(turn: HistoryTurn): CardElement[] {
  const out: CardElement[] = [];
  if (turn.userText.trim()) out.push(md(`**👤 你**\n${truncate(turn.userText, USER_MAX)}`));
  if (turn.assistantText.trim()) out.push(md(`**🤖 Codex**\n${truncate(turn.assistantText, ASSIST_MAX)}`));
  if (!turn.assistantText.trim() && !turn.userText.trim() && turn.tools.length) {
    out.push(noteMd('_（仅工具调用，无文本回复）_'));
  }

  const detail: CardElement[] = [];
  if (turn.reasoning.trim()) detail.push(md(`🧠 **思考**\n${truncate(turn.reasoning, REASON_MAX)}`));
  if (turn.tools.length) detail.push(md(toolsBlock(turn.tools)));
  if (detail.length) {
    out.push(collapsiblePanelEl({ title: detailTitle(turn), expanded: false, border: 'blue', elements: detail }));
  }
  return out;
}

function detailTitle(turn: HistoryTurn): string {
  const parts: string[] = [];
  if (turn.reasoning.trim()) parts.push('🧠 思考');
  if (turn.tools.length) parts.push(`🧰 ${turn.tools.length} 个工具`);
  return `🔎 ${parts.join(' · ')}`;
}

function toolsBlock(tools: HistoryTool[]): string {
  const lines: string[] = [`🧰 **工具调用（${tools.length}）**`];
  let body = 0;
  let shown = 0;
  for (const t of tools) {
    if (shown >= TOOLS_MAX_LINES || body >= TOOLS_BODY_MAX) {
      lines.push(`_…还有 ${tools.length - shown} 个_`);
      break;
    }
    const line = toolLine(t);
    lines.push(line);
    body += line.length;
    shown += 1;
  }
  return lines.join('\n');
}

function toolLine(t: HistoryTool): string {
  const title = escapeInline(truncate(oneLine(t.title), TOOL_TITLE_MAX));
  const mark = t.failed ? ' ✗' : '';
  const exit = t.exitCode != null && t.exitCode !== 0 ? ` (exit ${t.exitCode})` : '';
  return `- \`${title}\`${mark}${exit}`;
}

/** Rough rendered-content size of an element list, for the budget guard. */
function estimateSize(els: CardElement[]): number {
  let n = 0;
  for (const el of els) n += JSON.stringify(el).length;
  return n;
}

function oneLine(s: string): string {
  return s.replace(/\s+/g, ' ').trim();
}

/** Strip backticks so an inline-code span can't break out of its fence. */
function escapeInline(s: string): string {
  return s.replace(/`/g, '');
}

function truncate(s: string, n: number): string {
  const t = s.trim();
  return t.length > n ? `${t.slice(0, n)}…` : t;
}

/** Keep the END of `s`, prefixing '…' when content precedes it — used by
 * "上次停在" so it shows where the last reply actually stopped. */
function truncateTail(s: string, n: number): string {
  const t = oneLine(s);
  return t.length > n ? `…${t.slice(t.length - n)}` : t;
}

File: internal/card/history_card.go
package card

// history_card.go —— resume 历史卡（对齐 TS card/history-card）。
// 纯摘要：每轮一个折叠面板（Q 在标题、A 在体、思考/工具更深一层折叠）；18KB 预算从新到老分配。

import (
	"encoding/json"
	"fmt"
	"strings"
	"time"

	"github.com/modelzen/feishu-codex-bridge/internal/agent"
)

// 截断/预算常量（对齐 TS）。
const (
	histUserMax       = 300
	histAssistMax     = 800
	histReasonMax     = 600
	histToolTitleMax  = 90
	histToolsBodyMax  = 700
	histToolsMaxLines = 12
	histPreviewMax    = 160
	histTitleQMax     = 56
	panelShell        = 360
	histBodyBudget    = 18000
)

// HistoryCardState 历史卡状态。
type HistoryCardState struct {
	Cwd         string
	ProjectName string
	History     agent.ThreadHistory
}

// BuildHistoryCard 历史卡（now 用于 relativeTime）。
func BuildHistoryCard(s HistoryCardState, now time.Time) CardObject {
	elements := []CardElement{histMetaNote(s.Cwd, s.ProjectName)}
	if len(s.History.Turns) == 0 {
		elements = append(elements, Hr(), Md("_这个会话还没有可显示的历史（可能是空会话或刚创建）。_"), Hr(), resumedFooter())
		return Card(elements, CardOpts{Header: histHeader(s, now), Summary: "已恢复历史会话"})
	}
	dropped := s.History.TotalTurns - len(s.History.Turns)
	if dropped > 0 {
		elements = append(elements, Note(fmt.Sprintf("仅显示最近 %d 轮，更早的 %d 轮 Codex 仍保留在上下文中。", len(s.History.Turns), dropped)))
	}
	elements = append(elements, Hr())

	// 预算从新到老分配；emit 倒 oldest→newest。
	panels := []CardElement{}
	used := 0
	for i := len(s.History.Turns) - 1; i >= 0; i-- {
		turn := s.History.Turns[i]
		title := turnTitle(turn)
		body := turnBody(turn)
		size := estimateSize(body) + panelShell
		if used+size > histBodyBudget && len(panels) > 0 {
			stub := "_（内容已省略，历史较长）_"
			panels = append(panels, CollapsiblePanel(CollapsiblePanelOpts{Title: title, Expanded: false, Border: "grey", Body: stub}))
			used += len(title) + len(stub) + panelShell
		} else {
			panels = append(panels, CollapsiblePanelEl(title, false, "grey", body))
			used += size
		}
	}
	// reverse
	for i, j := 0, len(panels)-1; i < j; i, j = i+1, j-1 {
		panels[i], panels[j] = panels[j], panels[i]
	}
	elements = append(elements, panels...)

	last := s.History.Turns[len(s.History.Turns)-1]
	leftOff := ""
	if last.AssistantText != "" {
		leftOff = last.AssistantText
	} else if last.UserText != "" {
		leftOff = last.UserText
	} else if len(last.Tools) > 0 {
		leftOff = last.Tools[len(last.Tools)-1].Title
	}
	if strings.TrimSpace(leftOff) != "" {
		elements = append(elements, Hr(), NoteMd(fmt.Sprintf("📍 **上次停在**：%s", truncateTail(leftOff, histPreviewMax))))
	} else {
		elements = append(elements, Hr())
	}
	elements = append(elements, resumedFooter())
	return Card(elements, CardOpts{Header: histHeader(s, now), Summary: fmt.Sprintf("已恢复历史会话 · 共 %d 轮", s.History.TotalTurns)})
}

func histHeader(s HistoryCardState, now time.Time) *CardHeader {
	bits := []string{}
	name := strings.TrimSpace(s.History.Name)
	if name == "" {
		name = s.ProjectName
	}
	if name != "" {
		bits = append(bits, name)
	}
	bits = append(bits, fmt.Sprintf("共 %d 轮", s.History.TotalTurns))
	if s.History.UpdatedAt != 0 {
		bits = append(bits, RelativeTime(s.History.UpdatedAt, now))
	}
	return &CardHeader{Title: "🕘 已恢复历史会话", Template: HeaderTurquoise, Subtitle: strings.Join(bits, " · ")}
}

func histMetaNote(cwd, projectName string) CardElement {
	parts := []string{"📂 `" + cwd + "`"}
	if projectName != "" {
		return Note("📁 " + projectName + "   " + parts[0])
	}
	return Note(parts[0])
}

func resumedFooter() CardElement {
	return Md("✅ **会话已恢复** —— 直接发消息即可继续。")
}

func turnTitle(turn agent.HistoryTurn) string {
	if strings.TrimSpace(turn.UserText) != "" {
		return "👤 " + escapeInline(truncate(oneLine(turn.UserText), histTitleQMax))
	}
	return "⚙️ 系统 / 工具调用"
}

func turnBody(turn agent.HistoryTurn) []CardElement {
	var out []CardElement
	if strings.TrimSpace(turn.UserText) != "" {
		out = append(out, Md(fmt.Sprintf("**👤 你**\n%s", truncate(turn.UserText, histUserMax))))
	}
	if strings.TrimSpace(turn.AssistantText) != "" {
		out = append(out, Md(fmt.Sprintf("**🤖 Codex**\n%s", truncate(turn.AssistantText, histAssistMax))))
	}
	if strings.TrimSpace(turn.AssistantText) == "" && strings.TrimSpace(turn.UserText) == "" && len(turn.Tools) > 0 {
		out = append(out, NoteMd("_（仅工具调用，无文本回复）_"))
	}
	var detail []CardElement
	if strings.TrimSpace(turn.Reasoning) != "" {
		detail = append(detail, Md(fmt.Sprintf("🧠 **思考**\n%s", truncate(turn.Reasoning, histReasonMax))))
	}
	if len(turn.Tools) > 0 {
		detail = append(detail, Md(toolsBlock(turn.Tools)))
	}
	if len(detail) > 0 {
		out = append(out, CollapsiblePanelEl(detailTitle(turn), false, "blue", detail))
	}
	return out
}

func detailTitle(turn agent.HistoryTurn) string {
	parts := []string{}
	if strings.TrimSpace(turn.Reasoning) != "" {
		parts = append(parts, "🧠 思考")
	}
	if len(turn.Tools) > 0 {
		parts = append(parts, fmt.Sprintf("🧰 %d 个工具", len(turn.Tools)))
	}
	return "🔎 " + strings.Join(parts, " · ")
}

func toolsBlock(tools []agent.HistoryTool) string {
	lines := []string{fmt.Sprintf("🧰 **工具调用（%d）**", len(tools))}
	body := 0
	shown := 0
	for _, t := range tools {
		if shown >= histToolsMaxLines || body >= histToolsBodyMax {
			lines = append(lines, fmt.Sprintf("_…还有 %d 个_", len(tools)-shown))
			break
		}
		line := toolLine(t)
		lines = append(lines, line)
		body += len(line)
		shown++
	}
	return strings.Join(lines, "\n")
}

func toolLine(t agent.HistoryTool) string {
	title := escapeInline(truncate(oneLine(t.Title), histToolTitleMax))
	mark := ""
	if t.Failed {
		mark = " ✗"
	}
	exit := ""
	if t.ExitCode != nil && *t.ExitCode != 0 {
		exit = fmt.Sprintf(" (exit %d)", *t.ExitCode)
	}
	return fmt.Sprintf("- `%s`%s%s", title, mark, exit)
}

func estimateSize(els []CardElement) int {
	n := 0
	for _, el := range els {
		b, _ := json.Marshal(el)
		n += len(b)
	}
	return n
}

func oneLine(s string) string { return collapseSpaces(strings.TrimSpace(s)) }

func escapeInline(s string) string { return strings.ReplaceAll(s, "`", "") }

func truncateTail(s string, n int) string {
	t := oneLine(s)
	runes := []rune(t)
	if len(runes) > n {
		return "…" + string(runes[len(runes)-n:])
	}
	return t
}
